pkg/virt-launcher/notify-client/client.go

```diff
@@ -151,19 +154,25 @@ func libvirtEventCallback(c cli.Connection, domain *api.Domain, libvirtEvent lib
 	}
 }
 
-func (c *NotifyClient) StartDomainNotifier(domainConn cli.Connection, deleteNotificationSent chan watch.Event, vmiUID types.UID) error {
-
+func (c *NotifyClient) StartDomainNotifier(domainConn cli.Connection, deleteNotificationSent chan watch.Event, vmiUID types.UID, qemuAgentPollerInterval *time.Duration) error {
 	eventChan := make(chan libvirtEvent, 10)
 	reconnectChan := make(chan bool, 10)
 
 	domainConn.SetReconnectChan(reconnectChan)
 
+	qemuAgentCallback := func(domainName string, interfaceStatuses *[]api.InterfaceStatus) {
+		eventCallback(domainConn, util.NewDomainFromName(domainName, vmiUID), libvirtEvent{}, c, deleteNotificationSent, interfaceStatuses)
+	}
+	agentPoller := agentpoller.CreatePoller(domainConn, vmiUID, qemuAgentCallback, qemuAgentPollerInterval)
+
 	// Run the event process logic in a separate go-routine to not block libvirt
 	go func() {
 		for {
 			select {
```

Comment:
I would design it a little bit different:

 1) if we get a guest-agent connected event, we start a go-routine which periodically queries the guest agent
 2) if there is a value change from the last query, we emit an event in a channel where this switch statement watches for changes.
 3) if we get a disconnect event, we close a "done" channel of the started watcher and the watch stops querying
 4) If we get another "connected" event, and the agent is not already running create a new instance of the agent and start polling again.